test: Global Kill e2e test is flaky

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

Run [merged e2e tests](https://prow.tidb.net/?repo=pingcap%2Ftidb&type=postsubmit&job=pingcap%2Ftidb%2Fmerged_e2e_test).

### 2. What did you expect to see? (Required)

Test succeed.

### 3. What did you see instead (Required)

Test failed:

```
[2023/12/08 20:13:30.297 +08:00] [INFO] [global_kill_test.go:283] ["starting tidb"] [cmd="bin/globalkilltest_tidb-server --store=tikv -L=info --path=127.0.0.1:2379 -P=5001 --status=8001 --log-file=/tmp/tidb_globalkilltest/tidb5001.log --log-slow-query=/tmp/tidb_globalkilltest/tidb-slow5001.log --config=./config-64.toml"]
[2023/12/08 20:13:30.797 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=0] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:31.048 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=1] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:31.549 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=2] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:32.549 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=3] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:33.550 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=4] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:34.551 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=5] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:35.552 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=6] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:36.552 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=7] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:37.553 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=8] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:38.553 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=9] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:39.554 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=10] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:40.555 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=11] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:41.555 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=12] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:42.556 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=13] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:43.556 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=14] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:44.556 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=15] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:45.557 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=16] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:46.557 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=17] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:47.558 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=18] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:48.559 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=19] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:49.560 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=20] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:50.560 +08:00] [WARN] [global_kill_test.go:354] ["ping addr failed"] [addr=127.0.0.1:5001] ["retry count"=21] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:51.561 +08:00] [ERROR] [global_kill_test.go:367] ["connect to server addr failed"] [addr=127.0.0.1:5001] ["take time"=20.764030926s] [error="dial tcp 127.0.0.1:5001: connect: connection refused"]
[2023/12/08 20:13:51.561 +08:00] [INFO] [global_kill_test.go:299] ["stopping: bin/globalkilltest_tidb-server --store=tikv -L=info --path=127.0.0.1:2379 -P=5001 --status=8001 --log-file=/tmp/tidb_globalkilltest/tidb5001.log --log-slow-query=/tmp/tidb_globalkilltest/tidb-slow5001.log --config=./config-64.toml"]
[2023/12/08 20:13:51.571 +08:00] [INFO] [global_kill_test.go:301] ["stopped: bin/globalkilltest_tidb-server --store=tikv -L=info --path=127.0.0.1:2379 -P=5001 --status=8001 --log-file=/tmp/tidb_globalkilltest/tidb5001.log --log-slow-query=/tmp/tidb_globalkilltest/tidb-slow5001.log --config=./config-64.toml"]
[2023/12/08 20:13:51.815 +08:00] [INFO] [global_kill_test.go:243] ["cluster cleaned"]
--- FAIL: TestOneTiDB (32.53s)
    global_kill_test.go:554: 
        	Error Trace:	/home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb/tests/globalkilltest/global_kill_test.go:554
        	            				/home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb/tests/globalkilltest/global_kill_test.go:539
        	Error:      	Received unexpected error:
        	            	dial tcp 127.0.0.1:5001: connect: connection refused
        	            	github.com/pingcap/errors.AddStack
        	            		/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174
        	            	github.com/pingcap/errors.Trace
        	            		/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15
        	            	github.com/pingcap/tidb/tests/globalkilltest.(*GlobalKillSuite).connectTiDB
        	            		/home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb/tests/globalkilltest/global_kill_test.go:372
        	            	github.com/pingcap/tidb/tests/globalkilltest.doTestOneTiDB
        	            		/home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb/tests/globalkilltest/global_kill_test.go:553
        	            	github.com/pingcap/tidb/tests/globalkilltest.TestOneTiDB
        	            		/home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb/tests/globalkilltest/global_kill_test.go:539
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:1595
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1650
        	Test:       	TestOneTiDB
```

See https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fmerged_e2e_test/detail/merged_e2e_test/2987/pipeline/.

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

master (899dfe8a7417a545a0c049c7d77876c8eaee5667)

--- a/tests/globalkilltest/BUILD.bazel
+++ b/tests/globalkilltest/BUILD.bazel
@@ -1,4 +1,4 @@
-load("@io_bazel_rules_go//go:def.bzl", "go_test")
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_test(
     name = "globalkilltest_test",
@@ -7,6 +7,7 @@ go_test(
         "global_kill_test.go",
         "main_test.go",
     ],
+    embed = [":globalkilltest"],
     flaky = True,
     deps = [
         "//pkg/testkit/testsetup",
@@ -18,6 +19,20 @@ go_test(
         "@io_etcd_go_etcd_client_v3//:client",
         "@org_golang_google_grpc//:grpc",
         "@org_golang_google_grpc//backoff",
+        "@org_uber_go_zap//:zap",
+    ],
+)
+
+go_library(
+    name = "globalkilltest",
+    srcs = ["util.go"],
+    importpath = "github.com/pingcap/tidb/tests/globalkilltest",
+    visibility = ["//visibility:public"],
+    deps = [
+        "//pkg/server",
+        "//pkg/util",
+        "@com_github_pingcap_errors//:errors",
+        "@com_github_pingcap_log//:log",
         "@org_uber_go_zap//:zap",
     ],
 )

--- a/tests/globalkilltest/global_kill_test.go
+++ b/tests/globalkilltest/global_kill_test.go
@@ -170,8 +170,7 @@ func (s *GlobalKillSuite) startTiKV(dataDir string) (err error) {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	time.Sleep(500 * time.Millisecond)
-	return nil
+	return errors.Trace(CheckTiKVStatus())
 }
 
 func (s *GlobalKillSuite) startPD(dataDir string) (err error) {
@@ -185,8 +184,7 @@ func (s *GlobalKillSuite) startPD(dataDir string) (err error) {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	time.Sleep(500 * time.Millisecond)
-	return nil
+	return errors.Trace(CheckPDHealth(*pdClientPath))
 }
 
 func (s *GlobalKillSuite) startCluster() (err error) {
@@ -199,7 +197,6 @@ func (s *GlobalKillSuite) startCluster() (err error) {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	time.Sleep(10 * time.Second)
 	return nil
 }
 
@@ -266,8 +263,7 @@ func (s *GlobalKillSuite) startTiDBWithoutPD(port int, statusPort int) (cmd *exe
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	time.Sleep(500 * time.Millisecond)
-	return cmd, nil
+	return cmd, errors.Trace(CheckTiDBStatus(statusPort))
 }
 
 func (s *GlobalKillSuite) startTiDBWithPD(port int, statusPort int, pdPath string) (cmd *exec.Cmd, err error) {
@@ -285,8 +281,7 @@ func (s *GlobalKillSuite) startTiDBWithPD(port int, statusPort int, pdPath strin
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	time.Sleep(500 * time.Millisecond)
-	return cmd, nil
+	return cmd, errors.Trace(CheckTiDBStatus(statusPort))
 }
 
 func (s *GlobalKillSuite) mustStartTiDBWithPD(t *testing.T, port int, statusPort int, pdPath string) *exec.Cmd {
@@ -335,9 +330,8 @@ func (s *GlobalKillSuite) connectTiDB(port int) (db *sql.DB, err error) {
 	dsn := fmt.Sprintf("root@(%s)/test", addr)
 	sleepTime := 250 * time.Millisecond
 	sleepTimeLimit := 1 * time.Second
-	maxRetryDuration := 20 * time.Second
 	startTime := time.Now()
-	for i := 0; time.Since(startTime) < maxRetryDuration; i++ {
+	for i := 0; time.Since(startTime) < timeoutConnectDB; i++ {
 		db, err = sql.Open("mysql", dsn)
 		if err != nil {
 			log.Warn("open addr failed",

--- a/tests/globalkilltest/global_kill_test.go
+++ b/tests/globalkilltest/global_kill_test.go
@@ -170,7 +170,7 @@ func (s *GlobalKillSuite) startTiKV(dataDir string) (err error) {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	return errors.Trace(CheckTiKVStatus())
+	return errors.Trace(checkTiKVStatus())
 }
 
 func (s *GlobalKillSuite) startPD(dataDir string) (err error) {
@@ -184,7 +184,7 @@ func (s *GlobalKillSuite) startPD(dataDir string) (err error) {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	return errors.Trace(CheckPDHealth(*pdClientPath))
+	return errors.Trace(checkPDHealth(*pdClientPath))
 }
 
 func (s *GlobalKillSuite) startCluster() (err error) {
@@ -263,7 +263,7 @@ func (s *GlobalKillSuite) startTiDBWithoutPD(port int, statusPort int) (cmd *exe
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return cmd, errors.Trace(CheckTiDBStatus(statusPort))
+	return cmd, errors.Trace(checkTiDBStatus(statusPort))
 }
 
 func (s *GlobalKillSuite) startTiDBWithPD(port int, statusPort int, pdPath string) (cmd *exec.Cmd, err error) {
@@ -281,7 +281,7 @@ func (s *GlobalKillSuite) startTiDBWithPD(port int, statusPort int, pdPath strin
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return cmd, errors.Trace(CheckTiDBStatus(statusPort))
+	return cmd, errors.Trace(checkTiDBStatus(statusPort))
 }
 
 func (s *GlobalKillSuite) mustStartTiDBWithPD(t *testing.T, port int, statusPort int, pdPath string) *exec.Cmd {

--- a/tests/globalkilltest/global_kill_test.go
+++ b/tests/globalkilltest/global_kill_test.go
@@ -58,8 +58,9 @@ var (
 )
 
 const (
-	waitToStartup   = 500 * time.Millisecond
-	msgErrConnectPD = "connect PD err: %v. Establish a cluster with PD & TiKV, and provide PD client path by `--pd=<ip:port>[,<ip:port>]"
+	waitToStartup    = 500 * time.Millisecond
+	msgErrConnectPD  = "connect PD err: %v. Establish a cluster with PD & TiKV, and provide PD client path by `--pd=<ip:port>[,<ip:port>]"
+	timeoutConnectDB = 20 * time.Second
 )
 
 // GlobalKillSuite is used for automated test of "Global Kill" feature.

--- a/tests/globalkilltest/util.go
+++ b/tests/globalkilltest/util.go
@@ -1,0 +1,135 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package globalkilltest
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/pkg/server"
+	"github.com/pingcap/tidb/pkg/util"
+	"go.uber.org/zap"
+)
+
+const (
+	timeoutCheckPDStatus   = 10 * time.Second
+	timeoutCheckTiKVStatus = 30 * time.Second
+	timeoutCheckTiDBStatus = 30 * time.Second
+	timeoutConnectDB       = 20 * time.Second
+
+	retryInterval = 500 * time.Millisecond
+)
+
+func WithRetry[T any](fn func() (T, error), timeout time.Duration) (resp T, err error) {
+	startTime := time.Now()
+	retry := 0
+	for time.Since(startTime) < timeout {
+		retry += 1
+		resp, err = fn()
+		if err == nil {
+			break
+		}
+		log.Debug("withRetry", zap.Error(err), zap.Int("retry", retry))
+		time.Sleep(retryInterval)
+	}
+	return resp, errors.Trace(err)
+}
+
+func CheckPDHealth(host string) error {
+	url := util.ComposeURL(host, "/health")
+
+	request := func() (interface{}, error) {
+		resp, err := util.InternalHTTPClient().Get(url)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("PD health status code %d", resp.StatusCode)
+		}
+		defer resp.Body.Close()
+		var health struct {
+			Health string `json:"health"`
+		}
+		dec := json.NewDecoder(resp.Body)
+		err = dec.Decode(&health)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		log.Info("PD health", zap.Any("health", health))
+		if health.Health != "true" {
+			return nil, fmt.Errorf("PD not healthy %v", health.Health)
+		}
+		return nil, nil
+	}
+
+	_, err := WithRetry(request, timeoutCheckPDStatus)
+	return errors.Trace(err)
+}
+
+func CheckTiKVStatus() error {
+	url := util.ComposeURL("127.0.0.1:20180", "/status")
+
+	request := func() (interface{}, error) {
+		resp, err := util.InternalHTTPClient().Get(url)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		log.Info("TiKV status", zap.Any("status", resp.StatusCode))
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("TiKV status code %d", resp.StatusCode)
+		}
+		resp.Body.Close()
+		return nil, nil
+	}
+
+	_, err := WithRetry(request, timeoutCheckTiKVStatus)
+	return errors.Trace(err)
+}
+
+func CheckTiDBStatus(statusPort int) error {
+	host := fmt.Sprintf("127.0.0.1:%d", statusPort)
+	url := util.ComposeURL(host, "/status")
+
+	request := func() (interface{}, error) {
+		resp, err := util.InternalHTTPClient().Get(url)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("TiDB status code %d", resp.StatusCode)
+		}
+		defer resp.Body.Close()
+		var status server.Status
+		dec := json.NewDecoder(resp.Body)
+		err = dec.Decode(&status)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		log.Info("TiDB status", zap.Any("status", status))
+		return nil, nil
+	}
+
+	_, err := WithRetry(request, timeoutCheckTiDBStatus)
+	return errors.Trace(err)
+}

--- a/tests/globalkilltest/util.go
+++ b/tests/globalkilltest/util.go
@@ -30,13 +30,13 @@ import (
 const (
 	timeoutCheckPDStatus   = 10 * time.Second
 	timeoutCheckTiKVStatus = 30 * time.Second
-	timeoutCheckTiDBStatus = 30 * time.Second
+	timeoutCheckTiDBStatus = 60 * time.Second // First start up of TiDB would take a long time.
 	timeoutConnectDB       = 20 * time.Second
 
 	retryInterval = 500 * time.Millisecond
 )
 
-func WithRetry[T any](fn func() (T, error), timeout time.Duration) (resp T, err error) {
+func withRetry[T any](fn func() (T, error), timeout time.Duration) (resp T, err error) {
 	startTime := time.Now()
 	retry := 0
 	for time.Since(startTime) < timeout {
@@ -51,7 +51,7 @@ func WithRetry[T any](fn func() (T, error), timeout time.Duration) (resp T, err 
 	return resp, errors.Trace(err)
 }
 
-func CheckPDHealth(host string) error {
+func checkPDHealth(host string) error {
 	url := util.ComposeURL(host, "/health")
 
 	request := func() (interface{}, error) {
@@ -80,11 +80,11 @@ func CheckPDHealth(host string) error {
 		return nil, nil
 	}
 
-	_, err := WithRetry(request, timeoutCheckPDStatus)
+	_, err := withRetry(request, timeoutCheckPDStatus)
 	return errors.Trace(err)
 }
 
-func CheckTiKVStatus() error {
+func checkTiKVStatus() error {
 	url := util.ComposeURL("127.0.0.1:20180", "/status")
 
 	request := func() (interface{}, error) {
@@ -101,11 +101,11 @@ func CheckTiKVStatus() error {
 		return nil, nil
 	}
 
-	_, err := WithRetry(request, timeoutCheckTiKVStatus)
+	_, err := withRetry(request, timeoutCheckTiKVStatus)
 	return errors.Trace(err)
 }
 
-func CheckTiDBStatus(statusPort int) error {
+func checkTiDBStatus(statusPort int) error {
 	host := fmt.Sprintf("127.0.0.1:%d", statusPort)
 	url := util.ComposeURL(host, "/status")
 
@@ -130,6 +130,6 @@ func CheckTiDBStatus(statusPort int) error {
 		return nil, nil
 	}
 
-	_, err := WithRetry(request, timeoutCheckTiDBStatus)
+	_, err := withRetry(request, timeoutCheckTiDBStatus)
 	return errors.Trace(err)
 }

--- a/tests/globalkilltest/util.go
+++ b/tests/globalkilltest/util.go
@@ -31,7 +31,6 @@ const (
 	timeoutCheckPDStatus   = 10 * time.Second
 	timeoutCheckTiKVStatus = 30 * time.Second
 	timeoutCheckTiDBStatus = 60 * time.Second // First start up of TiDB would take a long time.
-	timeoutConnectDB       = 20 * time.Second
 
 	retryInterval = 500 * time.Millisecond
 )

